File: Excel/900-999/992/Challenge 992.py
import pandas as pd
from itertools import combinations

path = "900-999/992/992 Maximal Combinations.xlsx"
input = pd.read_excel(path, usecols="A:B", nrows=28, skiprows=1)
test = pd.read_excel(path, usecols="D:E", nrows=10, skiprows=1)

items_by_order = input.groupby("OrderID", sort=False)["Item"].apply(list)

result = (
	pd.DataFrame(
		(
			(",".join(sorted(map(str, c))), len(c))
			for items in items_by_order
			for k in range(2, min(4, len(items)) + 1)
			for c in combinations(items, k)
		),
		columns=["Combination", "length"],
	)
	.groupby(["Combination", "length"], as_index=False)
	.size()
	.rename(columns={"size": "Support"})
	.loc[lambda d: d["Support"] >= 2]
	.sort_values(["length", "Support", "Combination"], ascending=[False, False, True])
	.drop(columns="length")
	.reset_index(drop=True)
)

# Support identical, order of combination in test is inconsistent.

File: Excel/900-999/992/Challenge 992.R
library(tidyverse)
library(readxl)

path <- "900-999/992/992 Maximal Combinations.xlsx"
input <- read_excel(path, range = "A2:B29")
test <- read_excel(path, range = "D2:E12")

result = input %>%
  summarise(items = list(Item), .by = OrderID) %>%
  transmute(
    Combination = map(
      items,
      ~ flatten(map(2:min(4, length(.x)), \(k) combn(.x, k, simplify = FALSE)))
    )
  ) %>%
  unnest(Combination) %>%
  mutate(
    Combination = map_chr(Combination, ~ paste(sort(.x), collapse = ",")),
    length = str_count(Combination, ",") + 1
  ) %>%
  summarise(Support = n(), .by = c(Combination, length)) %>%
  filter(Support >= 2) %>%
  arrange(desc(length), desc(Support), Combination) %>%
  select(-length)

# Support identical, order of combination in test is incosistent.
